File: ideapad-perf-tray.rb
#!/usr/bin/ruby
require 'gtk3'

si=Gtk::StatusIcon.new
si.set_from_icon_name("gnome-power-manager")

#check active modes and set active menu option based on that

def vbc
  vbcheck = `ideapad-perf -vb`
  if vbcheck == "Rapid Charge mode is on, Battery Conservation mode is off.".chomp
    $vb = "rc"
  elsif vbcheck == "Rapid Charge mode is off, Battery Conservation mode is on.".chomp
    $vb = "bc"
  elsif vbcheck == "Rapid Charge mode is off, Battery Conservation mode is off.".chomp
    $vb = "off"
  end
end

#setting the performance and battery modes
def setbmode(x)
  `ideapad-perf -b #{x}`
end

#check initial states
vbc

#menu
menu=Gtk::Menu.new

group = nil

label2=Gtk::MenuItem.new(label: "Battery mode")
label2.set_sensitive(false)
menu.append(label2)

group2 = nil

item4=Gtk::RadioMenuItem.new(group2, "Rapid Charge")
menu.append(item4)
item4.active = true if $vb == "rc"
item4.signal_connect("activate") { 
  if item4.active? == true
    setbmode("rc")
  end
}

item5=Gtk::RadioMenuItem.new(item4, "Battery Conservation")
menu.append(item5)
item5.active = true if $vb == "bc"
item5.signal_connect("activate") { 
  if item5.active? == true
    setbmode("bc")
  end
}
 
item6=Gtk::RadioMenuItem.new(item4, "Off")
menu.append(item6)
item6.active = true if $vb == "off"
item6.signal_connect("activate") { 
  if item6.active? == true
    setbmode("off")
  end
}

menu.append(Gtk::SeparatorMenuItem.new)

quit=Gtk::MenuItem.new(label: 'Quit')
quit.signal_connect('activate'){Gtk.main_quit}
menu.append(quit)
quit.show
menu.show_all


##Show menu on right click
si.signal_connect('popup-menu'){|tray, button, time| menu.popup(nil, nil, button, time)}

Gtk.main

File: ideapad-perf-tray.py
#!/usr/bin/python

import signal
import subprocess

import gi

gi.require_version("Gtk", "3.0")
gi.require_version("AppIndicator3", "0.1")
from gi.repository import AppIndicator3 as appindicator
from gi.repository import Gtk as gtk


class Indicator:
    def __init__(self):
        self.vbc()
        self.filenames = {
            "rc": "/usr/share/icons/Papirus-Dark/symbolic/status/battery-good-charging-symbolic.svg",
            "bc": "/usr/share/icons/Papirus-Dark/symbolic/status/battery-good-symbolic.svg",
            "off": "/usr/share/icons/Papirus-Dark/symbolic/status/battery-missing-symbolic.svg",
        }
        self.icon_filename: str = self.filenames[self.battery_profile]
        self.indicator = appindicator.Indicator.new(
            "customtray",
            self.icon_filename,
            appindicator.IndicatorCategory.APPLICATION_STATUS,
        )
        self.indicator.set_status(appindicator.IndicatorStatus.ACTIVE)
        self.indicator.set_menu(self.menu())

    def vbc(self):
        f = subprocess.check_output("ideapad-perf -vb".split()).decode("utf-8").strip()
        match f:
            case "Rapid Charge mode is on, Battery Conservation mode is off.":
                self.battery_profile = 'rc'
            case "Rapid Charge mode is off, Battery Conservation mode is on.":
                self.battery_profile = 'bc'
            case "Rapid Charge mode is off, Battery Conservation mode is off.":
                self.battery_profile = 'off'

    def menu(self):
        menu = gtk.Menu()

        title = gtk.MenuItem(label="Battery mode")
        title.set_sensitive(False)
        menu.append(title)

        mode_1 = gtk.RadioMenuItem(label="Rapid charge")
        menu.append(mode_1)
        if self.battery_profile == "rc":
            mode_1.set_active(True)
        mode_1.connect("activate", self.change_mode, "rc")

        mode_2 = gtk.RadioMenuItem(label="Battery conservation", group=mode_1)
        if self.battery_profile == "bc":
            mode_2.set_active(True)
        menu.append(mode_2)
        mode_2.connect("activate", self.change_mode, "bc")

        mode_3 = gtk.RadioMenuItem(label="Off", group=mode_1)
        menu.append(mode_3)
        if self.battery_profile == "off":
            mode_3.set_active(True)
        mode_3.connect("activate", self.change_mode, "off")

        exit_tray = gtk.MenuItem(label="Exit Tray")
        menu.append(exit_tray)
        exit_tray.connect("activate", self.quit, "quit")

        menu.show_all()
        return menu

    def change_mode(self, source, string):
        subprocess.call(f"ideapad-perf -b {string}".split(), stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)
        self.vbc()
        if self.battery_profile == string:
            self.indicator.set_icon_full(self.filenames[string], "")

    def quit(*args):
        gtk.main_quit(*args)


Indicator()
signal.signal(signal.SIGINT, signal.SIG_DFL)
gtk.main()
